src/cluster.ml

```diff
@@ -0,0 +1,213 @@
+  archs : Arch.t list;
+}
+
+type service = [
+(* Services on deploy.ci.dev *)
+  | `Ci of string
+  | `Opamrepo of string
+  | `Check of string
+  | `Ci3 of string
+  | `Ci4 of string
+  | `Docs of string
+  | `Staging_docs of string
+
+  (* Services on deploy.mirage.io *)
+  | `Cimirage of string
+
+  (* Services on deploy.ci.ocaml.org. *)
+  | `Ocamlorg_deployer of string             (* OCurrent deployer @ deploy.ci.ocaml.org *)
+  | `OCamlorg_v2 of (string * string option) list   (* OCaml website @ v2.ocaml.org *)
+  | `Ocamlorg_images of string               (* Base Image builder @ images.ci.ocaml.org *)
+  | `OCamlorg_v3b of string                  (* OCaml website @ v3b.ocaml.org aka www.ocaml.org *)
+  | `OCamlorg_v3c of string                  (* Staging OCaml website @ staging.ocaml.org *)
+  | `Aws_ecs of Aws.t                        (* Amazon Web Services - Elastic Container Service *)
+]
+
+type deploy_info = {
+  hub_id : Cluster_api.Docker.Image_id.t;
+  services : service list;
+}
+
+let get_job_id x =
+  let+ md = Current.Analysis.metadata x in
+  match md with
+  | Some { Current.Metadata.job_id; _ } -> job_id
+  | None -> None
+
+let unwrap = function
+  | `Path _ as x -> Current.return x
+  | `Contents x -> Current.map (fun x -> `Contents x) x
+
+let component_label label dockerfile pool =
+  let pp_label = Fmt.(option (cut ++ string)) in
+  match dockerfile with
+  | `Path path -> Current.component "build %s@,%s%a" path pool pp_label label
+  | `Contents _ -> Current.component "build@,%s%a" pool pp_label label
+
+let ocluster_build ?level ?label ?cache_hint t ~pool ~src ~options dockerfile =
+  component_label label dockerfile pool |>
+  let> dockerfile = unwrap dockerfile
+  and> options
+  and> src in
+  Current_ocluster.Raw.build ?level ?cache_hint t ~pool ~src ~options dockerfile
+
+(* Build [src/dockerfile] as a Docker service. *)
+let build { sched; dockerfile; options; archs } ?(additional_build_args=Current.return []) repo src =
+  let options =
+    let+ additional_build_args = additional_build_args in
+    { options with build_args = additional_build_args @ options.build_args }
+  in
+  let hash = Current.map Current_git.Commit_id.hash src in
+  let build_arch arch =
+    let src = Current.map (fun x -> [x]) src in
+    let pool = Arch.pool_id arch in
+    let build = ocluster_build sched ~options ~pool ~src dockerfile in
+    let index =
+      let+ job_id = get_job_id build
+      and+ hash in
+      let label =
+        match dockerfile with
+        | `Path path -> Fmt.str "build %s@,%s" path pool
+        | `Contents _ -> Fmt.str "build@,%s" pool
+      in
+      Index.record ~repo ~hash [(label, job_id)]
+    in
+    Current.all [build; index]
+  in
+  Current.all (List.map build_arch archs)
+
+let name info = Cluster_api.Docker.Image_id.to_string info.hub_id
+
+let no_schedule = Current_cache.Schedule.v ()
+
+let pull_and_serve (module D : Current_docker.S.DOCKER) ~name op repo_id =
```

Comment:
Since `D` is a first class module anyhow, and since we do not use any of the modules defined for this anywhere else, I think we could ditch lines 36 to 51 entirely by tweaking this function to be something like:

```suggestion
let pull_and_serve ?docker_context ~name op repo_id =
  let module D : Current_docker.S.DOCKER =
    if Option.is_some docker_context then
      Current_docker.Default
    else
      Current_docker.Make(struct let docker_context = docker_context end)
  in
```

Reply:
Ah, I may be mistaken here, as I see we do use at least one of these submodules at https://github.com/shonfeder/ocurrent-deployer/blob/9bc4cb9be497145ae3763d93d0e081287fa4e2f7/src/pipeline.ml#L583